nexus/src/db/explain.rs

```diff
@@ -0,0 +1,239 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Utility allowing Diesel to EXPLAIN queries.
+
+use super::pool::DbConnection;
+use async_bb8_diesel::{AsyncRunQueryDsl, ConnectionManager, PoolError};
+use async_trait::async_trait;
+use diesel::pg::Pg;
+use diesel::prelude::*;
+use diesel::query_builder::*;
+
+/// A wrapper around a runnable Diesel query, which EXPLAINs what it is doing.
+///
+/// Q: The Query we're explaining.
+///
+/// EXPLAIN: https://www.cockroachlabs.com/docs/stable/explain.html
+pub trait Explainable<Q> {
+    /// Syncronously issues an explain statement.
+    fn explain(
+        self,
+        conn: &mut DbConnection,
+    ) -> Result<String, diesel::result::Error>;
+}
+
+impl<Q> Explainable<Q> for Q
+where
+    Q: QueryFragment<Pg> + RunQueryDsl<DbConnection> + Sized,
+{
+    fn explain(
+        self,
+        conn: &mut DbConnection,
+    ) -> Result<String, diesel::result::Error> {
+        Ok(ExplainStatement { query: self }
+            .get_results::<String>(conn)?
+            .join("\n"))
+    }
+}
+
+/// An async variant of [`Explainable`].
+#[async_trait]
+pub trait ExplainableAsync<Q> {
+    /// Asynchronously issues an explain statement.
+    async fn explain_async(
+        self,
+        pool: &bb8::Pool<ConnectionManager<DbConnection>>,
+    ) -> Result<String, PoolError>;
+}
+
+#[async_trait]
+impl<Q> ExplainableAsync<Q> for Q
+where
+    Q: QueryFragment<Pg> + RunQueryDsl<DbConnection> + Sized + Send + 'static,
+{
+    async fn explain_async(
+        self,
+        pool: &bb8::Pool<ConnectionManager<DbConnection>>,
+    ) -> Result<String, PoolError> {
+        Ok(ExplainStatement { query: self }
+            .get_results_async::<String>(pool)
+            .await?
+            .join("\n"))
+    }
+}
+
+// An EXPLAIN statement, wrapping an underlying query.
+//
+// This isn't `pub` because it's kinda weird to access "part" of the EXPLAIN
+// output, which would be possible by calling "get_result" instead of
+// "get_results". We'd like to be able to constrain callers such that they get
+// all of the output or none of it.
+//
+// See the [`Explainable`] trait for why this exists.
+struct ExplainStatement<Q> {
+    query: Q,
+}
+
+impl<Q> QueryId for ExplainStatement<Q> {
+    type QueryId = ();
+    const HAS_STATIC_QUERY_ID: bool = false;
```

Comment:
Would it make sense to derive this based on `Q` or is this `false` because we don't really expect to use these with prepared statements and hence don't need to worry about caching them?

Reply:
Good question... My experience with prepared statements is a little limited, so if you have a strong opinion, lemme know!

I do think that optimizing for `EXPLAIN` to be used in prepared statements is a little... weird? Like, these should be one-offs, I *think*, but maybe I'm wrong about that.

To propagate the `HAS_QUERY_STATIC_ID` value, I could use:
```
  const HAS_STATIC_QUERY_ID: bool = Q::HAS_STATIC_QUERY_ID;
```

But I'd need to make a new type to represent the "EXPLAIN" wrapper when setting `type QueryId = ...`, right?

As in, I couldn't just do:

```rust
  type QueryId = Q::QueryId;
```

Because that would potentially conflict with the underlying query type.

Frankly: I'm not 100% certain about my usage here, and picking "false" felt like the conservative choice.

Reply:
> To propagate the `HAS_QUERY_STATIC_ID` value, I could use:
> 
> ```rust
>   const HAS_STATIC_QUERY_ID: bool = Q::HAS_STATIC_QUERY_ID;
> ```
> 
> But I'd need to make a new type to represent the "EXPLAIN" wrapper when setting `type QueryId = ...`, right?
> 
> As in, I couldn't just do:
> 
> ```rust
>   type QueryId = Q::QueryId;
> ```

Do you need a new type? You already have `ExplainStatement`. `#[derive(QueryId)]` would basically give you something like:

```rust
impl<Q> QueryId for ExplainStatement<Q>
where
    Q: QueryId,
{
    type QueryId = ExplainStatement<Q::QueryId>;

    const HAS_STATIC_QUERY_ID: bool = Q::HAS_STATIC_QUERY_ID;
}
```

>I do think that optimizing for EXPLAIN to be used in prepared statements is a little... weird? Like, these should be one-offs, I think, but maybe I'm wrong about that.

But yea I agree. I'm also not super familiar with how diesel even caches prepared statements.

Reply:
This is a good idea - I've added this implementation.